components/locid/src/langid.rs

```diff
@@ -154,14 +154,48 @@ impl std::fmt::Display for LanguageIdentifier {
     }
 }
 
+macro_rules! subtag_matches {
+    ($T:ty, $iter:ident, $expected:expr) => {
+        match $iter.next() {
+            Some(ref text) => {
+                let parsed: Result<$T, _> = text.parse();
+                match parsed {
+                    Ok(subtag) => subtag == $expected,
+                    Err(_) => false,
+                }
+            }
+            None => false,
+        }
+    };
+}
+
 impl PartialEq<&str> for LanguageIdentifier {
     fn eq(&self, other: &&str) -> bool {
-        self.to_string().eq(*other)
+        let mut iter = other.split('-').peekable();
```

Comment:
Also, we support `_` as a subtag separator. In fact, we now have 4 places in this component where we do that, so I'd suggest you add a function in `parser/mod.rs` like:

```rust
pub fn get_subtag_iterator(t: &[u8]) -> impl Iterator<Item = &[u8]> {
     t.split(|c| *c == b'-' || *c == b'_')
}
```

and replace two calls in `extensions`, one in `locale` and one in `langid` with it and use it here.

For that to work, you'll need to change your macro to use `from_bytes` instead of `parse`:

```rust
macro_rules! subtag_matches {
    ($T:ty, $iter:ident, $expected:expr) => {
        match $iter.next() {
            Some(ref b) => {
                match <$T>::from_bytes(b) {
                    Ok(subtag) => subtag == $expected,
                    Err(_) => false,
                }
            }
            None => false,
        }
    };
}
```

Reply:
If you want to make your macro a biit more rusty, you can also use the chaining:

```rust
        $iter
            .next()
            .map(|b| <$T>::from_bytes(b) == Ok($expected))
            .unwrap_or(false)
```

It doesn't impact disassembled code tho, so whatever is more readable in your opinion.